src/main/javatests/com/google/summit/translation/SourceLocationTest.kt

```diff
@@ -116,12 +115,38 @@ class SourceLocationTest {
             = 'Hello';
         }
         """.trim()
-    val loc = SourceLocation(startLine = 1, startColumn = 4, endLine = 3, endColumn = 16)
+    val loc = SourceLocation(startLine = 1, startColumn = 4, endLine = 3, endColumn = 15)
 
     val extracted = loc.extractFrom(input)
 
     assertThat(extracted).isEqualTo("""s Test {
           public String field
             = 'H""")
   }
+
+  @Test
+  fun trailingEmptyLine() {
+    val input = "public class Test { }\n"
+
+    val cu = TranslateHelpers.parseAndTranslate(input)
+    val loc = cu.getSourceLocation()
+
+    assertThat(loc.endLine).isEqualTo(2)
+    // There is no first column on the second, empty line.
+    // Expect zero in this case.
+    assertThat(loc.endColumn).isEqualTo(0)
+    assertThat(loc.extractFrom(input)).isEqualTo(input)
+  }
+
+  @Test
+  fun leadingEmptyLine() {
+    val input = "\npublic class Test { }"
+
+    val cu = TranslateHelpers.parseAndTranslate(input)
+    val loc = cu.getSourceLocation()
+
+    // The source location starts from the first regular token.
```

Comment:
why this lack of symmetry? And what happens if

val input = "\n public class Test { }"

or

val input = " public class Test { }"

?

Reply:
It follows from the grammar, not from my code that's representing source locations. No great reason. I threw in this test to see if there was also a degenerate case with `beginColumn` on an empty line, but I don't think it can happen with this grammar.

Reply:
(All of these cases have a `beginColumn` at the `p` character.)